File: WeatherApp-with-Angular/src/app/body/weather.service.ts
import { HttpClient } from '@angular/common/http';
import { Injectable } from '@angular/core';
import { Observable } from 'rxjs';

import { apiKey } from '../shared/constants/appConstants';

import IResponseData from '../shared/interfaces/IResponseData';
import IGeneralData from '../shared/interfaces/IGeneralData';
import IBodyData from '../shared/interfaces/IBodyData';

@Injectable()
export class WeatherService {
  isLoaded = false;

  constructor(private http: HttpClient) {}

  getWeatherData(cityName: string): Observable<IResponseData> {
    const apiURL =
      "https://api.openweathermap.org/data/2.5/weather?q=" +
      cityName +
      "&appid=" +
      apiKey;

    return this.http.get<IResponseData>(apiURL);
  }

  getGeneralData(data: IResponseData): IGeneralData {
    return {
      imgUrl: `http://openweathermap.org/img/wn/${data.weather[0].icon}@2x.png`,
      name: data.name,
      country: data.sys.country,
      description: data.weather[0].description,
      temp: ~~(data.main.temp - 273.15).toFixed(2),
      feels_like: ~~(data.main.feels_like - 273.15).toFixed(2) - 1
    }
  }

  getBodyData(data: IResponseData): IBodyData[] {
    const pressure: IBodyData = {
      icon: 'fa-solid fa-gauge-high',
      value: data?.main?.pressure,
      units: 'hPa',
      text: 'Pressure'
    };
    const humidity: IBodyData = {
      icon: 'fa-solid fa-droplet',
      value: data?.main?.humidity,
      units: '%',
      text: 'Humidity'
    };
    const visibility: IBodyData = {
      icon: 'fa-solid fa-eye',
      value: data?.visibility / 1000,
      units: 'km',
      text: 'Visibility'
    };
    const windSpeed: IBodyData = {
      icon: 'fa-solid fa-wind',
      value: data?.wind?.speed,
      units: 'km/h',
      text: 'Wind Speed'
    };
    const windDirection: IBodyData = {
      icon: 'fa-solid fa-compass',
      value: data?.wind?.deg,
      units: 'deg',
      text: 'Wind Direction'
    };
    const cloudiness: IBodyData = {
      icon: 'fa-solid fa-cloud',
      value: data?.clouds?.all,
      units: '%',
      text: 'Cloudiness'
    };

    const bodyData: IBodyData[] = [
      pressure, 
      humidity, 
      visibility, 
      windSpeed, 
      windDirection, 
      cloudiness
    ];

    return bodyData;
  }
}

File: WeatherApp-with-ASP/WeatherApp.Core/Services/WeatherService.cs
using WeatherApp.Core.Contracts;
using Newtonsoft.Json;
using Microsoft.Extensions.Configuration;
using WeatherApp.Core.Models.DTO;
using WeatherApp.Core.Models.ViewModels;

namespace WeatherApp.Core.Services
{
    public class WeatherService : IWeatherService
    {
        private readonly HttpClient _client;

        private readonly string _apiKey;
        private readonly string _baseUrl;

        public WeatherService(HttpClient client, IConfiguration configuration)
        {
            _client = client;
            _apiKey = configuration["ApiConstants:ApiKey"];
            _baseUrl = configuration["ApiConstants:BaseUrl"];
        }

        public async Task<Tuple<GeneralInfoViewModel, List<InfoCardViewModel>>> GetWeatherInfoAsync(string cityName)
        {
            var url = _baseUrl + "weather?q=" + cityName + "&appid=" + _apiKey;

            string json = await _client.GetStringAsync(url);
            WeatherInfoModel weatherInfo = JsonConvert.DeserializeObject<WeatherInfoModel>(json);

            var generalInfo = CreateGeneralInfo(weatherInfo);
            var infoCards = CreateInfoCards(weatherInfo);

            return new Tuple<GeneralInfoViewModel, List<InfoCardViewModel>>(generalInfo, infoCards);
        }

        private List<InfoCardViewModel> CreateInfoCards(WeatherInfoModel weatherInfo)
        {
            var pressure = new InfoCardViewModel
            {
                Icon = "fa-solid fa-gauge-high",
                Value = weatherInfo.Main.Pressure,
                Units = "hPa",
                Text = "Pressure"
            };

            var humidity = new InfoCardViewModel
            {
                Icon = "fa-solid fa-droplet",
                Value = weatherInfo.Main.Humidity,
                Units = "%",
                Text = "Humidity"
            };

            var visibility = new InfoCardViewModel
            {
                Icon = "fa-solid fa-eye",
                Value = weatherInfo.Visibility / 1000,
                Units = "km",
                Text = "Visibility"
            };

            var windSpeed = new InfoCardViewModel
            {
                Icon = "fa-solid fa-wind",
                Value = weatherInfo.Wind.Speed,
                Units = "km/h",
                Text = "Wind Speed"
            };

            var windDirection = new InfoCardViewModel
            {
                Icon = "fa-solid fa-compass",
                Value = weatherInfo.Wind.Deg,
                Units = "deg",
                Text = "Wind Direction"
            };

            var cloudiness = new InfoCardViewModel
            {
                Icon = "fa-solid fa-cloud",
                Value = weatherInfo.Clouds.All,
                Units = "%",
                Text = "Cloudiness"
            };

            return new List<InfoCardViewModel>
            {
                pressure,
                humidity,
                visibility,
                windSpeed,
                windDirection,
                cloudiness
            };
        }

        private GeneralInfoViewModel CreateGeneralInfo (WeatherInfoModel weatherInfo)
        {
            var generalInfo = new GeneralInfoViewModel
            {
                Icon = weatherInfo.Weather[0].Icon,
                Name = weatherInfo.Name,
                Country = weatherInfo.Sys.Country,
                Description = weatherInfo.Weather[0].Description,
                Temp = Math.Floor(weatherInfo.Main.Temp - 273.15).ToString("N0"),
                FeelsLike = Math.Floor(weatherInfo.Main.Feels_like - 273.15 - 1).ToString("N0")
            };

            return generalInfo;
        }
    }
}
